### PR TITLE
Add support for Active Technologies AWG-401x

### PR DESCRIPTION
**Pull request by garzetti**
Active Technologies AWG-401x is a dual mode Arbitrary Function
Generator (AFG) and Arbitrary Waveform Generator (AWG) 1.2GS/s
instrument. This instrument is available with two (AWG-4012), four
(AWG-4014), and eight (AWG-4018) channel configurations, all supported
by this module. Depending on the model, 8, 16, or 32 digital channels
are available as options.

This module implements the interface of both modes in two distinct
classes: AWG401x_AFG and AWG401x_AWG.

In AWG401x_AFG, most of the functionality of AFG mode has been
implemented: selection of waveshape, amplitude, frequency, offset,
phase, ... . All the parameters are validated at runtime based on the
validity ranges provided by the instrument, which may change based on
hardware configuration and license status.

In AWG401x_AWG, the basic functionality of AWG mode has been
implemented. This includes the selection of basic amplitude and time
parameters and the selection of the entries' waveforms. Additionally,
the possibility to upload custom waveforms has been included.

The module has been implemented using the official documentation
available at https://www.activetechnologies.it. Portions of the
documentation have been included in the methods' docstring after
written permission from Active Technologies.

Notable missing features (currently) are the control of digital
channels, the flow control of the entries in AWG mode, and the chaining
of multiple instruments.

### CODE REVIEW

File: AUTHORS.txt [hunk 1]
```diff
@@ -49,3 +49,4 @@ Jack Van Sambeek
 JC Arbelbide
 Florian Jünger
 Benedikt Moneke
+Fabio Garzetti
```
Comment (msmttchr) [inline AUTHORS.txt:52]: Please update also docs/about/authors.rst.
Unfortunately we have this duplication.

File: pymeasure/instruments/activetechnologies/AWG401x.py [hunk 2]
```diff
@@ -0,0 +1,1145 @@
...
+        get_process=lambda v: AWG401x_AWG._run_state_dict[v]
+    )
+
+    enabled = Instrument.control(
+        "AWGControl:RSTATe?", "AWGControl:%s",
+        """A boolean property that enables the generation of signals.""",
+        validator=strict_discrete_set,
+        values={True: "RUN", False: "STOP"},
+        map_values=True,
+        get_process=lambda v: "STOP" if v == 0 else "RUN"
```
Comment (msmttchr) [inline pymeasure/instruments/activetechnologies/AWG401x.py:692]: This property has different commands to set and get. Is this intentional?

**Review by msmttchr — CHANGES_REQUESTED** [review]
@garzetti, first batch of comments on this very interesting PR. I like you tried to address problems like instrument channels with dynamic properties.

File: pymeasure/instruments/activetechnologies/AWG401x.py [hunk 3]
```diff
@@ -0,0 +1,1145 @@
...
+        ValueError
+            If position is outside permitted limit [0,4].
+        """
+
+        if position >= 0 or position <= 4:
+            self.write(f"*RCL {position}")
+        else:
+            raise ValueError("position value outside permitted range [0,4]")
+
+    def waitLast(self):
```
[the remark this answers is not in the document]
Comment (msmttchr) [reply pymeasure/instruments/activetechnologies/AWG401x.py:453]: Similar comment for other method's name

File: pymeasure/instruments/activetechnologies/AWG401x.py [hunk 4]
```diff
@@ -0,0 +1,1145 @@
...
+        "AFGControl:STATus?", "AFGControl:%s",
+        """A boolean property that enables the generation of signals.""",
+        validator=strict_discrete_set,
+        values={True: "START", False: "STOP"},
+        map_values=True,
+        get_process=lambda v: "START" if v == 1 else ("STOP" if v == 0 else v)
+    )
+
+    def __init__(self, adapter, **kwargs):
+        super().__init__(adapter)
```
Comment (msmttchr) [inline pymeasure/instruments/activetechnologies/AWG401x.py:498]: kwargs are silently discarded. Is this intentional?

File: pymeasure/instruments/activetechnologies/AWG401x.py [hunk 5]
```diff
@@ -0,0 +1,1145 @@
...
+            self.voltage_high_values = list(map(
+                float,
+                [self.voltage_high_max, self.voltage_high_min]))
+
+            self.voltage_low_values = list(map(
+                float,
+                [self.voltage_low_max, self.voltage_low_min]))
+
+
+class IEEE488DataBlock():
```
Comment (msmttchr) [inline pymeasure/instruments/activetechnologies/AWG401x.py:1112]: I wonder if this class can be removed by using utilities provided by `pyvisa`.
See pymeasure/adapters/serial.py for some example

Comment (physicsninja) [reply pymeasure/instruments/activetechnologies/AWG401x.py:1112]: @msmttchr For some context, this instrument can also accept data in the form of a text string of numbers separated by newlines, not just arrays of numbers. When I wrote a much more hacky implementation of this instrument (under the US distributer name Berkeley Nucleonics 675)  [here](https://github.com/neal-kepler/pymeasure/blob/master/pymeasure/instruments/berkeleynucleonics/bn675.py), I made the same choice. In light of your comment I think [write_binary](https://pyvisa.readthedocs.io/en/latest/introduction/rvalues.html) with datatype='s' may actually work, I didn't realize at the time that you could pass raw strings. @garzetti can correct me if they have already tried.

File: pymeasure/instruments/activetechnologies/AWG401x.py [hunk 6]
```diff
@@ -0,0 +1,1145 @@
...
+    )
+
+    polarity = Instrument.control(
+        "OUTPut<ch>:POLarity?", "OUTPut<ch>:POLarity %s",
+        """This property inverts the output waveform relative to its average
+        value: (High Level – Low Level)/2. NORM for normal, INV for inverted
+        """,
+        validator=strict_discrete_set,
+        values=["NORMAL", "NORM", "INVERTED", "INV"],
+        get_process=lambda v: "NORM" if v == 0 else ("INV" if v == 1 else v)
```
[the remark this answers is not in the document]
Comment (garzetti) [reply pymeasure/instruments/activetechnologies/AWG401x.py:62]: Because the get and set arguments are different. The set use `NORMAL` and `INVERTED` while get use `0` and `1`

File: pymeasure/instruments/activetechnologies/AWG401x.py [hunk 7]
```diff
@@ -0,0 +1,1145 @@
...
+    )
+
+    delay = Instrument.control(
+        None, None,
```
[the remark this answers is not in the document]
Comment (garzetti) [reply pymeasure/instruments/activetechnologies/AWG401x.py:66]: Yes, since the command for the AFG and AWG channel's delay are different i will set in the dedicated class derived from `Channel_base`. e.g., https://github.com/pymeasure/pymeasure/blob/9e0d254463e05259eff6093f21b7c761fc5aba42/pymeasure/instruments/activetechnologies/AWG401x.py#L145-L149

File: pymeasure/instruments/activetechnologies/AWG401x.py [hunk 8]
```diff
@@ -0,0 +1,1145 @@
...
+        """This property queries the minimum delay that can be set to the
+        output waveform.""",
+        dynamic=True
+    )
+
+    def __init__(self, instrument, channel_number):
+        self.instrument = instrument
+        self.channel_number = channel_number
+
+        self._special_names = self._setup_special_names()
```
[the remark this answers is not in the document]
Comment (garzetti) [reply pymeasure/instruments/activetechnologies/AWG401x.py:95]: Yes, right

File: pymeasure/instruments/activetechnologies/AWG401x.py (hunk 2 above)
Comment (garzetti) [reply pymeasure/instruments/activetechnologies/AWG401x.py:692]: Yes, it is.  Here there are tree different command for a single property:
 * AWGControl: RSTATe? (Query Only)
 * AWGControl:RUN (No Query Form)
 * AWGControl:STOP (No Query Form)

File: AUTHORS.txt (hunk 1 above)
Comment (garzetti) [reply AUTHORS.txt:52]: Done

File: pymeasure/instruments/activetechnologies/AWG401x.py [hunk 9]
```diff
@@ -0,0 +1,1145 @@
...
+
+    def __init__(self, adapter, **kwargs):
+
+        try:
+            if isinstance(adapter, (int, str)):
+                adapter = VISAAdapter(adapter, **kwargs)
+                # Insert an higher timeout because often, when starting the
+                # instrument, can pass some time
+                adapter.connection.timeout = 7500
+        except ImportError:
```
[the remark this answers is not in the document]
Comment (garzetti) [reply pymeasure/instruments/activetechnologies/AWG401x.py]: Done

File: pymeasure/instruments/activetechnologies/AWG401x.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (garzetti) [reply pymeasure/instruments/activetechnologies/AWG401x.py]: Done

File: pymeasure/instruments/activetechnologies/AWG401x.py [hunk 10]
```diff
@@ -0,0 +1,1145 @@
...
+        """This property queries the minimum sample rate for the Sampling
+        Clock."""
+    )
+
+    _run_state_dict = {0: "STOPPED", 1: "WAITING_TRIGGER", 2: "RUNNING"}
+
+    run_status = Instrument.measurement(
+        "AWGControl:RSTATe?",
+        """This property returns the run state of the AWG. The possible values
+        are: STOPPED, WAITING_TRIGGER, RUNNING""",
```
[the remark this answers is not in the document]
Comment (garzetti) [reply pymeasure/instruments/activetechnologies/AWG401x.py:674]: Done

File: pymeasure/instruments/activetechnologies/AWG401x.py (hunk 4 above)
Comment (garzetti) [reply pymeasure/instruments/activetechnologies/AWG401x.py]: Fixed

File: pymeasure/instruments/activetechnologies/AWG401x.py (hunk 5 above)
Comment (garzetti) [reply pymeasure/instruments/activetechnologies/AWG401x.py]: I have removed the `IEEE488DataBlock` class, as suggested by @msmttchr , and used `from_ieee_block` and `to_ieee_block` methods from pyvisa.util . 
As suggested by @physicsninja , probably using `query_binary_values` and `write_binary_values` from pyvisa will work, but in my opinion, it is better to modify the `pymeasure.adapter.visa` class to include these two methods instead of calling it directly from pyvisa.

**Comment by garzetti** [comment]
Feel free to squash the commits at the end of PR review process

File: pymeasure/instruments/activetechnologies/AWG401x.py (hunk 5 above)
Comment (msmttchr) [reply pymeasure/instruments/activetechnologies/AWG401x.py]: What about passing the keyword parameter `datatype='s'` to `binary_values` and `write_binary_values` ? Does this the trick?

Comment (garzetti) [reply pymeasure/instruments/activetechnologies/AWG401x.py]: The `write_binary_values` will work because it simply call the pyvisa `write_binary_values` method in which specifying `datatype='s'` works correctly. I can not use `binary_values` because it does not expose directly the pyvisa `query_binary_values` in which i must use `header_fmt='ieee'` and `datatype='d'` (yes, the data type are different read direction).
I can suggest two solution:
 * Use directly the `connection` object in visa adapter (i.e., `self.instrument.connection.query_binary_values(...)` )
 * Modify the visa adapter to expose directly a `query_binary_values` method like the `write_binary_values` one

Comment (msmttchr) [reply pymeasure/instruments/activetechnologies/AWG401x.py]: @physicsninja and @garzetti thanks for your valuable feedback and clarification.
Let me share my opinion below.

> The `write_binary_values` will work because it simply call the pyvisa `write_binary_values` method in which specifying `datatype='s'` works correctly. I can not use `binary_values` because it does not expose directly the pyvisa `query_binary_values` in which i must use `header_fmt='ieee'` and `datatype='d'` (yes, the data type are different read direction). I can suggest two solution:
> 
>     * Use directly the `connection` object in visa adapter (i.e., `self.instrument.connection.query_binary_values(...)` )
> 
This has the disadvantage to restrict support (when using GPIB interface) to adapters supported by pyvisa (e.g. cheap Prologix is not supported), but the instrument comes by default with LAN interface, so it should be OK.

>     * Modify the visa adapter to expose directly a `query_binary_values` method like the `write_binary_values` one
The adapter are under restructuring, so, I would avoid to add more methods, especially because this needs to be kept consistent with other adapters. I would be in favor of doing this extension in a dedicated PR. If you have time have a look at the ongoing discussion in #617.

In summary, I would vote for the first option you suggested.

File: pymeasure/instruments/activetechnologies/AWG401x.py [hunk 11]
```diff
@@ -0,0 +1,1106 @@
...
+
+    def __init__(self, adapter, **kwargs):
+        super().__init__(adapter, **kwargs)
+
+        self.setting_ch = {}
+        for i in range(1, self.num_ch+1):
+            self.setting_ch[i] = ChannelAWG(self, i)
+
+        self.entries = self.DummyEntriesElements(self, self.num_ch)
+        # self.sequencer = Sequencer(self, self.num_ch)
```
Comment (msmttchr) [inline pymeasure/instruments/activetechnologies/AWG401x.py:715]:
```suggestion
```

File: pymeasure/instruments/activetechnologies/AWG401x.py [hunk 12]
```diff
@@ -0,0 +1,1106 @@
...
+
+    def trigger(self):
+        """Force a trigger event to occour."""
+        self.write("TRIGger:SEQuence:IMMediate")
+
+    def save_file(self,
+                  file_name: str,
+                  data,
+                  path: str = None,
+                  override_existing=False):
```
Comment (msmttchr) [inline pymeasure/instruments/activetechnologies/AWG401x.py:734]: type hinting is not encouraged in `PyMeasure`, at least from some discussion dated back in 2020 (see #259).

File: pymeasure/instruments/activetechnologies/AWG401x.py (hunk 5 above)
Comment (garzetti) [reply pymeasure/instruments/activetechnologies/AWG401x.py]: @msmttchr thanks for your suggestion. Modified to use the first solution.

**Comment by garzetti** [comment]
In the commit 2c618132d86e11383795cff38130902e07f8dfb1 I have modified how the properties are converted in float. Now the string are converted directly via `get_process` lambda. @msmttchr let me know if you prefer as it was before

File: pymeasure/instruments/activetechnologies/AWG401x.py (hunk 12 above)
Comment (garzetti) [reply pymeasure/instruments/activetechnologies/AWG401x.py]: type hinting removed

**Comment by msmttchr** [comment]
> In the commit [2c61813](https://github.com/pymeasure/pymeasure/commit/2c618132d86e11383795cff38130902e07f8dfb1) I have modified how the properties are converted in float. Now the string are converted directly via `get_process` lambda. @msmttchr let me know if you prefer as it was before

Sorry, I have missed this point during my original review. Actually, measurements properties cast everything to `float` by default and you can override this by passing a keyword parameter 'cast=<something>'. This is achieved because the property `fget` calls `values` method in the adapter and it passes `cast` parameter to it.
A bit weird for my taste.
Long story to say that there's no need to add extra code to cast to `float`.

**Comment by garzetti** [comment]
> > In the commit [2c61813](https://github.com/pymeasure/pymeasure/commit/2c618132d86e11383795cff38130902e07f8dfb1) I have modified how the properties are converted in float. Now the string are converted directly via `get_process` lambda. @msmttchr let me know if you prefer as it was before
> 
> Sorry, I have missed this point during my original review. Actually, measurements properties cast everything to `float` by default and you can override this by passing a keyword parameter 'cast='. This is achieved because the property `fget` calls `values` method in the adapter and it passes `cast` parameter to it. A bit weird for my taste. Long story to say that there's no need to add extra code to cast to `float`.

Sorry, I have not seen this cast in `values` function. I have removed every float cast in properties

**Review by msmttchr — APPROVED** [review]
Looks good to me.
Many thanks for your high quality contribution
